File: Programmers/Lv. 2/피로도.py
import itertools

def solution(k, dungeons):    
    l = len(dungeons)
    t = list(itertools.permutations([i for i in range(l)]))
    
    answer = -1
    
    for tt in t:
        c = 0
        temp = k
        for ttt in tt:
            if dungeons[ttt][0] <= temp:
                c += 1
                temp -= dungeons[ttt][1]
            else:
                break
        answer = max(answer, c)                        
    
    return answer

File: Programmers/Lv. 2/피로도.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int cnt = 0;

void dfs(vector<vector<int>> dungeons, bool *visited, int cur_h, int depth)
{
    if (depth == dungeons.size())
    {
        cnt = depth;
        return;
    }
    if (cur_h < 0)
    {
        cnt = max({cnt, depth-1});
        return;
    }
    
    for (int i = 0; i < dungeons.size(); i++)
    {
        if (!visited[i])
        {            
            visited[i] = true;
            
            int new_h;
            if (cur_h >= dungeons[i][0])
            {                
                new_h = cur_h - dungeons[i][1];  
                
                dfs(dungeons, visited, new_h, depth+1);
            }
            else
            {
                cnt = max({cnt, depth});
            }                
            visited[i] = false;
        }
    }
}

int solution(int k, vector<vector<int>> dungeons) {
    bool visited[8] = {false, };
    
    dfs(dungeons, visited, k, 0);
    
    return cnt;
}